File: src/java/org/javaturk/dp/ch05/adapter/electricity/problem2/TurkishHomeApplianceCompatibleWithUSPowerSource.java
package org.javaturk.dp.ch05.adapter.electricity.problem2;

import org.javaturk.dp.ch05.adapter.electricity.domain.tr.HomeAppliance;
import org.javaturk.dp.ch05.adapter.electricity.domain.tr.TurkishHomeAppliance;
import org.javaturk.dp.ch05.adapter.electricity.domain.tr.TurkishPowerSource;
import org.javaturk.dp.ch05.adapter.electricity.domain.us.USPowerSource;

/**
 * This handles the problem with extending the TurkishHomeAppliance.
 * @author akin
 *
 */
public class TurkishHomeApplianceCompatibleWithUSPowerSource extends TurkishHomeAppliance {
	private USPowerSource usPowerSource;
	private boolean turkishPowerSource;

	public TurkishHomeApplianceCompatibleWithUSPowerSource(String name) {
		super(name);
	}

	@Override
	public void setPowerSource(TurkishPowerSource powerSource) {
		this.powerSource = powerSource;
		turkishPowerSource = true;
	}

	public void setUSPowerSource(USPowerSource usPowerSource) {
		this.usPowerSource = usPowerSource;
		turkishPowerSource = false;
	}

	@Override
	public void start() {
		System.out.println("TurkishHomeAppliance " + name + " is starting!");
		if (turkishPowerSource)
			powerSource.turnOn();
		else
			usPowerSource.pushSwitch();
	}

	@Override
	public void stop() {
		System.out.println("TurkishHomeAppliance " + name + " stoping!");
		if (turkishPowerSource)
			powerSource.turnOff();
		else
			usPowerSource.pushSwitch();
	}
}

File: src/typescript/com/gokselkucuksahin/dp/structural-patterns/adapter/electricity/problem2/turkish-home-appliance-compatible-with-us-power-source.ts
import {TurkishHomeAppliance} from "../domain/tr/turkish-home-appliance";
import {USPowerSource} from "../domain/us/us-power-source";
import {TurkishPowerSource} from "../domain/tr/turkish-power-source";

export class TurkishHomeApplianceCompatibleWithUsPowerSource extends TurkishHomeAppliance {
  private usPowerSource: USPowerSource;
  private turkishPowerSource: boolean;

  constructor(name: string) {
    super(name);
  }


  setPowerSource(powerSource: TurkishPowerSource) {
    this.powerSource = powerSource;
    this.turkishPowerSource = true;
  }

  setUSPowerSource(usPowerSource: USPowerSource): void {
    this.usPowerSource = usPowerSource;
    this.turkishPowerSource = false;
  }


  start() {
    console.log(`TurkishHomeAppliance ${this.name} is starting!`);
    if (this.turkishPowerSource) {
      this.powerSource.turnOn();
    } else {
      this.usPowerSource.pushSwitch();
    }
  }

  stop() {
    console.log(`TurkishHomeAppliance ${this.name} is stoping!`);
    if (this.turkishPowerSource) {
      this.powerSource.turnOff();
    } else {
      this.usPowerSource.pushSwitch();
    }
  }
}